### PR TITLE
FillForward enumerator does not loop infinitely if fillforward resolution is different from data resolution

### LINKED ISSUE #5116: FillForward enumerator infinite loop
#### Expected Behavior

- FillForward enumerator does not loop infinitely
#### Actual Behavior

- FillForward enumerator is looping infinitely and blocking the backtest. Started happening with https://github.com/QuantConnect/Lean/pull/4941
#### Potential Solution

N/A
#### Reproducing the Problem

```
from datetime import timedelta

class BasicTemplateAlgorithm(QCAlgorithm):
    def Initialize(self):
        self.SetStartDate(2007, 1, 1)
        self.AddEquity('QQQ', Resolution.Hour)
        self.AddForex('EURUSD', Resolution.Daily, Market.FXCM)
```
#### System Information

N/A
#### Checklist

- [x] I have completely filled out this template
- [x] I have confirmed that this issue exists on the current `master` branch
- [x] I have confirmed that this is not a duplicate issue by searching [issues](https://github.com/QuantConnect/Lean/issues)

- [x] I have provided detailed steps to reproduce the issue

### PR DESCRIPTION
**Pull request by AdalyatNazirov**
#### Description

Calculate `potentialBarEndTime` in data time zone. `next.EndTime` is also replaced with `next.Time' converted into data time zone.

It's worth to consider this PR as continuation of https://github.com/QuantConnect/Lean/pull/4941.  There are two things that were not fixed in original PR.

During reading and parsing data from file we consider them as data in data time zone, and then convert them into exchange time zone. `TradeBar.EndTime` swallow Daylight Saving Time hour because is built using simple arithmetic operations (+/-).

That's why comparison `potentialBarEndTime` and `next.EndTime` isn't allowed - they can be in different TZ. To solve this problem we convert and compare dates built on top of `Start Time`.

Another change is that  `fillforward.EndTime` became fixed, and `fillforward.Time` is calculated by rewinding bar size. 

#### Related Issue

Closes #5116 

#### Motivation and Context

FillForward enumerator is looping infinitely and blocking the backtest. Started happening with https://github.com/QuantConnect/Lean/pull/4941

#### Requires Documentation Change

N/A

#### How Has This Been Tested?

More unit tests. Created unit tests around of Daylight Saving time changes for different fillforward and data resolutions.

#### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. 
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: Engine/DataFeeds/Enumerators/FillForwardEnumerator.cs [hunk 1]
```diff
@@ -300,7 +300,8 @@ protected virtual bool RequiresFillForwardData(TimeSpan fillForwardResolution, B
                 // it means that we have to calculate potentialBadEndTime in the same TZ
                 // by adding interval in daily resolution we move to next day
                 // in order to track daylight movement need to round-down origin day (before + Interval)
-                var potentialBarEndTime = RoundDown(item.ReferenceDateTime, item.Interval) + item.Interval;
+                var potentialBarEndTime = (RoundDown(item.ReferenceDateTime, item.Interval).ConvertTo(Exchange.TimeZone, _dataTimeZone) + item.Interval)
```
Comment (Martin-Molinero) [inline Engine/DataFeeds/Enumerators/FillForwardEnumerator.cs:303]: Believe we should update the comments above and the _why_ we are doing these operations

**Review by jaredbroad — COMMENTED** [review]
Pending 2 unit tests demonstrating the "skip" or "repeat" of the 2am Sunday bar during a timezone change. Expect the 2 am to not repeat due to how LEAN ignore <= bars; but should definitely see it _skip_ the bar when time moves forward.

**Comment by mchandschuh** [comment]
FYI -- time comparisons should _always_ use end time

**Review by Martin-Molinero — APPROVED** [review]
Thanks for all gray matter put into this one 👍

**Pull request merged**